group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java

```diff
@@ -135,56 +148,71 @@ private GroupAssignment newAssignmentHomogeneous(
+
+        // Combine current and new hashed assignments, sized to accommodate the expected number of mappings.
+        Map<String, Set<TopicIdPartition>> finalAssignment = new HashMap<>((int) ((numGroupMembers / 0.75f) + 1));
+        Map<TopicIdPartition, Set<String>> finalAssignmentByPartition = new HashMap<>((int) (((numTargetPartitions + 1) / 0.75f) + 1));
+
+        // First, take the members assigned by hashing.
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member -> {
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+            finalAssignmentByPartition.computeIfAbsent(targetPartition, k -> new HashSet<>()).add(member);
+        }));
 
-        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
-            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
+        // Then, take the members from the current assignment, making sure that no member has too many assigned partitions.
         // When combining current assignment, we need to only consider the topics in current assignment that are also being
         // subscribed in the new assignment as well.
         currentAssignment.forEach((targetPartition, members) -> {
-            if (subscribedTopicIds.contains(targetPartition.topicId()))
+            if (subscribedTopicIds.contains(targetPartition.topicId())) {
                 members.forEach(member -> {
-                    if (groupSpec.memberIds().contains(member) && !newAssignment.containsKey(targetPartition))
-                        finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+                    if (groupSpec.memberIds().contains(member)) {
+                        Set<TopicIdPartition> memberPartitions = finalAssignment.computeIfAbsent(member, k -> new HashSet<>());
+                        if ((memberPartitions.size() < desiredAssignmentCount) && !newAssignment.containsKey(targetPartition)) {
+                            memberPartitions.add(targetPartition);
+                            finalAssignmentByPartition.computeIfAbsent(targetPartition, k -> new HashSet<>()).add(member);
+                        }
+                    }
                 });
+            }
         });
 
+        // Finally, round-robin assignment for unassigned partitions which do not already have members assigned.
+        // The order of steps differs slightly from KIP-932 because the desired assignment count has been taken into
+        // account when copying partitions across from the current assignment, and this is more convenient.
+        List<TopicIdPartition> unassignedPartitions = targetPartitions.stream()
+            .filter(targetPartition -> !finalAssignmentByPartition.containsKey(targetPartition))
+            .toList();
+
+        roundRobinAssignmentWithCount(groupSpec.memberIds(), unassignedPartitions, finalAssignment, desiredAssignmentCount);
```

Comment:
Even though it should never be the case (since we calculate `desiredAssignmentCount` above), should we check that unassignedPartitions is empty after this call?

Reply:
It doesn't actually get modified. The intended safety check ensuring that we have a member that can take another partition. The exception is thrown if this is not true. I've added a test.